### PR TITLE
Reschedule too early events

### PR DESCRIPTION
**Pull request by netcriptus**
Solves #1149

### CODE REVIEW

File: tests/monitoring/monitoring_service/test_end_to_end.py [hunk 1]
```diff
@@ -143,6 +144,107 @@ def test_first_allowed_monitoring(
...
+    assert scheduled_events[0].trigger_timestamp == monitor_trigger
+
+    # Calling monitor too early must fail
+    monitoring_service.get_timestamp_now = lambda: settleable_after
+    monitoring_service._trigger_scheduled_events()  # pylint: disable=protected-access
+
+    # Failed event is rescheduled to run on the next iteration
+    scheduled_events = monitoring_service.database.get_scheduled_events(settleable_after)
+    assert len(scheduled_events) == 1
+    assert scheduled_events[0].trigger_timestamp != monitor_trigger
```
Comment (weilbith) [inline tests/monitoring/monitoring_service/test_end_to_end.py:245]:
```suggestion
    assert scheduled_events[0].trigger_timestamp > monitor_trigger
```

**Review by weilbith — COMMENTED** [review]
Just some nitpicks and a bigger discussion point (there no approval yet). But no critique to your work. Very good job. Simple and effective. 👍🏾 
Not your fault: I hate the end-to-end test. It is incredibly long with 90% basic stuff the exact case doesn't care about (which comes at the very end).

File: src/monitoring_service/service.py [hunk 2]
```diff
@@ -85,6 +87,14 @@ def handle_event(event: Event, context: Context) -> None:
                     "Processed event",
                     num_scheduled_events=context.database.scheduled_event_count(),
                 )
+            except TransactionTooEarlyException:
+                # When events are executed too early, they are rescheduled for retry.
+                log.debug("Event executed too early", handled_event=event)
+                context.database.upsert_scheduled_event(
+                    ScheduledEvent(
+                        trigger_timestamp=int(datetime.utcnow().timestamp()), event=event
```
[the remark this answers is not in the document]
Comment (weilbith) [reply src/monitoring_service/service.py:95]: We decided to open a new issue (#1153) for this and handle it independent from this PR here.

File: src/monitoring_service/handlers.py [hunk 3]
```diff
@@ -532,6 +533,8 @@ def action_monitoring_triggered_event_handler(event: Event, context: Context) ->
             )
         )
     except Exception as exc:  # pylint: disable=broad-except
+        if isinstance(exc, TransactionFailed) and "not allowed to monitor" in exc.args[0]:
```
[the remark this answers is not in the document]
Comment (netcriptus) [reply src/monitoring_service/handlers.py]: ✅

File: src/monitoring_service/service.py [hunk 4]
```diff
@@ -85,6 +87,14 @@ def handle_event(event: Event, context: Context) -> None:
                     "Processed event",
                     num_scheduled_events=context.database.scheduled_event_count(),
                 )
+            except TransactionTooEarlyException:
+                # When events are executed too early, they are rescheduled for retry.
```
[the remark this answers is not in the document]
Comment (netcriptus) [reply src/monitoring_service/service.py]: ✅

File: src/monitoring_service/service.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (netcriptus) [reply src/monitoring_service/service.py]: I decided to keep it simple and just set it to 10 seconds in the future. This should be safe enough to keep it under the timeout and long enough to not spam the node.

**Review by weilbith — APPROVED** [review]
👌🏾

**Pull request merged**